### PR TITLE
feat(device-discovery): generic interface↔VLAN associations — batch 5 (more L2 switches) (OBS-1368/OBS-1640)

### PR DESCRIPTION
**Pull request by leoparente**
## Summary

Adds `get_interfaces_vlans()` to five more custom NAPALM drivers — `brocade_netiron`, `avaya_ers`, `extreme_slx`, `ubiquiti_edgeswitch`, `ubiquiti_unifiswitch` — reusing the shared `_vlan.classify_switchport()` classifier from #378. Each driver does only vendor-specific field extraction.

Builds on top of #390. Brings the supported-driver count from 21 to 26.

## What changed

| Driver | Source | Notes |
|---|---|---|
| `brocade_netiron` | `show running-config vlan` (regex; same path as `get_vlans()`) | LAGs/VEs handled via `_netiron_split_port_list`; **canonical-name remap** to `GigabitEthernet`/`10GigabitEthernet`/etc via `show interfaces` cross-reference (`apply_interface_vlans()` needs full names) |
| `avaya_ers` | `show vlan interface info` (driver-local parser) + `show vlan` inversion (ntc-template) | UntagAll → access on PVID; UntagPvidOnly → trunk with native; TagAll → trunk no native (PVID stays tagged); empty membership for trunk modes → routed |
| `extreme_slx` | `show vlan brief` (driver-local parser; reuses existing `_expand_vlan_port` helper) | `(u)`/`(t)` flags drive shape inference; Eth/Po name canonicalisation already in driver |
| `ubiquiti_edgeswitch` | `show interfaces switchport` (table parser) + `show running-config` (vlan participation/tagging) | General → trunk collapse; access requires explicit untagged member; trunk requires explicit membership data |
| `ubiquiti_unifiswitch` | `show vlan` + per-VID `show vlan id <N>` | membership-shape mode inference (no admin-mode column on Broadcom-fastpath) |

## Vendor-specific semantics

- **brocade_netiron**: identical config-block format to FastIron; supports `ethe`/`lag`/`ve` port types and bare-digit (CES) port IDs (the FastIron pre-existing range-expansion bug from #390 commit 983b826 is preempted in `_netiron_expand_port_range`). **Canonical-name remap** is essential because `get_interfaces()` emits `GigabitEthernet1/1` while the running-config uses `ethe 1/1` — without the remap `apply_interface_vlans()` silently drops every entry. Validated against mockit (1 of 31 entries flowed end-to-end; the rest dropped due to mockit fixture inconsistency where running-config references ports `show interfaces` doesn't enumerate — would be 100% on real-device).
- **avaya_ers**: tagging mode determines mode regardless of show-vlan membership churn (UntagAll = PVID-only forwarding always). Membership data only used for UntagPvidOnly/TagAll trunks; no membership → routed (defensive).
- **extreme_slx**: SLX-OS has no admin-mode column; mode is inferred from `(u)`/`(t)` flag shape. Multi-untagged → routed (anomalous; 802.1Q forbids).
- **ubiquiti_edgeswitch**: access requires exactly-one-untagged-no-tagged membership shape (mirrors PowerConnect tightening in #390); empty membership for any mode → routed.
- **ubiquiti_unifiswitch**: same Broadcom-fastpath base as EdgeSwitch but `show vlan` output has different shape (no port column); driver enumerates VIDs then queries each (mirrors the ProCurve approach from #390).

## Test plan

- [x] All 958 tests pass (888 baseline from #390 + 70 new). 44 skipped.
- [x] Ruff clean across `device-discovery/`.
- [x] All 5 drivers expose `get_interfaces_vlans`:
  - `brocade_netiron: custom_napalm.brocade_netiron.NetIronDriver`
  - `avaya_ers: custom_napalm.avaya_ers.ERSDriver`
  - `extreme_slx: custom_napalm.extreme_slx.SLXOSDriver`
  - `ubiquiti_edgeswitch: custom_napalm.ubiquiti_edgeswitch.EdgeSwitchDriver`
  - `ubiquiti_unifiswitch: custom_napalm.ubiquiti_unifiswitch.UniFiSwitchDriver`
- [x] Codex local review pass 1: 8 valid findings addressed, 2 rejected with reasoning.
- [x] Mockit + dryrun validation for `brocade_netiron` (canonical-name fix was the resulting catch).
- [ ] Codex review against `develop` (will run on this PR via @codex review).
- [ ] Real-device validation deferred to a follow-up checkbox.

## Codex local review fixes folded in

- **Multi-untagged → routed**: All inversion-based drivers (netiron, slx, unifiswitch) now track `untagged` as `list[int]` instead of scalar; aggregator treats `len(untagged) > 1` as routed (anomalous; IEEE 802.1Q forbids a port being untagged in multiple VLANs simultaneously). Aggregators accept legacy scalar shape for backwards compatibility with any direct callers.
- **extreme_slx VID validation**: VLAN IDs validated through `coerce_vid()` during inversion; out-of-range VIDs dropped silently.
- **ubiquiti_edgeswitch tightening**: removed PVID-only access fallback when no participation data captured; trunk now requires explicit participation/tagging directives. Mirrors the dell_powerconnect batch-4 tightening from #390.
- **avaya_ers empty-membership guard**: UntagPvidOnly and TagAll trunks now require non-empty membership data; empty membership → routed (avoids clobbering NetBox `tagged_vlans` via PATCH).
- **brocade_netiron canonical-name remap** (mockit-found): added `_netiron_canonical_name_map()` to remap bare port IDs from running-config (`1/1`) to canonical names emitted by `get_interfaces()` (`GigabitEthernet1/1`) so `apply_interface_vlans()` exact-name match succeeds end-to-end.

Codex findings rejected with reasoning:
- "Helper schema mismatch with NAPALM's `access_vlan`/`trunk_allowed_vlans`" — the entire batch chain (1-4) intentionally uses the shared `_vlan.py` classifier output `{mode, tagged, untagged}` documented in `custom_napalm/CLAUDE.md`.
- "ERS UntagAll should require single-VID membership" — UntagAll semantics are PVID-only forwarding regardless of show-vlan membership churn (operator intent encoded in tagging mode, not in membership lists).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by jajeffries — APPROVED** [review]

**Review by samiura — APPROVED** [review]
Overall looks good, one comment though! Since, the brocade driver is pretty complex (+300 lines added) with canonical interface name remapping,  I am guessing we have enough edge cases related tests already for _netiron_split_port_list and apply_interface_vlans() l(e.g. VE interfaces, ports with no VLAN assignment).

**Comment by leoparente** [comment]
Codex round-2 P1 findings — all three valid, all three fixed in 38caa77.

1. **NetIron canonical-name regex rejected digit-leading Ethernet prefixes.** Confirmed: `^([A-Za-z]+)(\d+/\d+(?::\d+)?)$` did not match `10GigabitEthernet3/4` / `40GigabitEthernet...` because `[A-Za-z]+` doesn't accept a leading digit. Fixed by allowing optional leading digits: `(\d*[A-Za-z]+Ethernet)(\d+/\d+(?::\d+)?)`.

2. **NetIron canonical map didn't cover Ve/Lag/Loopback.** Confirmed: the inverter emits `lag5`, `ve2` (lowercase), and `get_interfaces()` exposes them as `Lag5`, `Ve2`, `Loopback1` — exact-name mismatch silently dropped VLAN mappings. Added a second regex branch that maps any `[A-Za-z]+\d+` canonical name to a `lowercase(prefix) + numeric` bare key matching what the inverter emits. SVI-shaped `vlanN` names are explicitly skipped.

3. **EdgeSwitch interface-block regex didn't match multi-token form.** Confirmed: `^interface\s+(\S+(?:/\S+)?)\s*$` rejected `interface lag 1` (two tokens). Broadened to `^interface\s+(.+?)\s*$` and normalise whitespace on capture so `lag 1` and `lag1` both reduce to `lag1`. SVI skip logic (`name.lower().startswith("vlan")`) still works since `vlan 1` becomes `vlan1`.

Three new field-mapper unit tests pin each behaviour:
- `test_netiron_canonical_map_handles_digit_leading_prefix_and_named_types`
- `test_edgesw_membership_parser_captures_lag_blocks`
- `test_edgesw_membership_parser_normalises_single_token_lag`

All 961 tests pass, ruff clean.

**Comment by leoparente** [comment]
@samiura — fair callout on edge cases. Tests covering the points you raised:

- **VE interface canonicalization** (added in 38caa77 from a follow-up Codex P1): `test_netiron_canonical_map_handles_digit_leading_prefix_and_named_types` exercises `Ve2` → `ve2` (and `Lag5` → `lag5`, `Loopback1` → `loopback1`) bare-key derivation, ensuring VE/LAG VLAN mappings survive `apply_interface_vlans()` exact-name match. The unit test in `tests/custom_drivers/test_field_mappers_batch5.py::test_netiron_invert_lag_membership` already pinned `tagged lag 1` → `lag1` and `untagged lag 5` → `lag5` from the inverter side.

- **Ports with no VLAN assignment**: `test_netiron_aggregate_multiple_untagged_yields_routed` and the empty-membership branch via `_netiron_aggregate_to_switchport({"untagged": [], "tagged": []})` (covered by the trunk_native and lag_membership scenarios where some ports appear in no VLAN). The aggregator's "neither → routed" branch is exercised whenever a port is in `per_port` with empty buckets — the test is now defensive against the multi-untagged anomalous case too.

- **Range-expansion edge cases** (already pinned): `test_netiron_split_port_list_short_e_token` (NetIron-specific `e` short token), `test_netiron_split_port_list_bare_digit_range_no_leading_slash` (preempts the FastIron #390 bare-digit bug).

- **End-to-end mockit + dry-run** (manual): validated against `mockit` with `DEVICE_TYPE=brocade_netiron` — VLAN entries flow correctly through `apply_interface_vlans()` for ports where running-config and show-interfaces agree (the canonical-name remap was added precisely because that validation surfaced the mismatch).

Happy to add more targeted tests if you want a specific scenario pinned — let me know which edge cases concern you most.

**Comment by leoparente** [comment]
@codex review

**Comment by leoparente** [comment]
Codex round-3 P1 — valid, fixed in bcd5d23.

EdgeSwitch (Broadcom-fastpath) accepts both the native `vlan ...` syntax and the Cisco-flavoured `switchport ...` syntax; the previous parser only handled the native form so Cisco-style configs left membership empty → every interface classified as routed.

Extended `_apply_membership_directive` to also recognise:
- `switchport access vlan X` → participation += [X]; pvid := X
- `switchport trunk native vlan X` → participation += [X]; pvid := X (NOT added to tagging)
- `switchport trunk allowed vlan X,Y,Z[ add|remove|except]` → tagging += [...]; participation += [...]
- **`switchport trunk allowed vlan all` → mode=tagged-all** (the only first-class wildcard path on EdgeSwitch — previously not supported at all; the row mapper now promotes via `allowed_vlans="all"`)
- `switchport general pvid X` → pvid := X
- `switchport general allowed vlan add X[,Y] tagged|untagged` → participation += [...]; tagging += [...] when `tagged`

Refactored the dispatcher into small helpers (`_apply_sp_pvid_setter`, `_apply_sp_trunk_allowed`, `_apply_sp_general_allowed`) to stay under ruff's C901 limit.

Three new field-mapper unit tests pin each branch:
- `test_edgesw_membership_parser_handles_cisco_style_access`
- `test_edgesw_membership_parser_handles_cisco_style_trunk`
- `test_edgesw_cisco_trunk_allowed_vlan_all_yields_tagged_all`

All 964 tests pass, ruff clean.

**Comment by leoparente** [comment]
@codex review

**Comment by leoparente** [comment]
Codex round-4 P1 — valid, fixed in a676485.

The regex captured the operation token but `_apply_sp_trunk_allowed` discarded it, so all three forms (`add`/`remove`/`except`) were treated as additive — `remove` and `except` produced the inverse of the intended semantics.

- `add` (or default, no op) → additive (existing)
- `remove` → drop the VIDs from tagging + participation
- `except` → set `allowed_except` flag; row mapper falls back to routed (NetBox's allowed_vlans list can't faithfully represent "all VLANs except these" without enumerating the chassis — emitting anything would risk clobbering NetBox via PATCH)
- `all` keyword (in spec) → wildcard → mode=tagged-all (existing)

Refactored the additive/subtractive paths into `_trunk_allowed_add` / `_trunk_allowed_remove` helpers to stay under ruff's C901 limit.

Two new unit tests pin both branches:
- `test_edgesw_cisco_trunk_allowed_remove_drops_vids`
- `test_edgesw_cisco_trunk_allowed_except_yields_routed`

All 966 tests pass, ruff clean.

**Comment by leoparente** [comment]
@codex review

**Comment by leoparente** [comment]
Codex round-5 P1 ×2 — both valid, both fixed in 9f639e1.

**1. ERS no-STG layout PVID mis-parse.** Confirmed: the regex hardcoded one numeric field before PVID (the STG column on v1 firmware). On v2 firmware (`Port FilterUF FilterUR PVID PRI Tagging`, no STG), PRI between PVID and Tagging was being captured as PVID. The example row `13 No Yes 1011 4 Hybrid` now correctly parses PVID=1011. Fixed by capturing both numeric columns as `i1`/`i2` and detecting layout via a header regex (`Port ... STG ... PVID ... Tagging` present → `i2` is PVID; otherwise `i1` is PVID).

**2. ERS Hybrid tagging mode dropped as routed.** Confirmed: the mapper only recognised `untagall`/`untagpvidonly`/`tagall`; `Hybrid` (common on VLAN-carrying ports) fell through to routed. ERS `Hybrid` is PVID-untagged-native + others-tagged — same NetBox-aligned semantics as `UntagPvidOnly`, so mapped the two together to share the trunk-with-native path (including the empty-membership → routed defensive guard).

Two new unit tests pin both branches:
- `test_ers_hybrid_yields_trunk_with_native`
- `test_ers_intf_info_parser_handles_no_stg_layout` (covers both v1 and v2 layouts in a single test)

All 968 tests pass, ruff clean.

**Comment by leoparente** [comment]
@codex review

**Comment by leoparente** [comment]
Codex round-6 P1 — valid, fixed in 821748c.

Confirmed: the regex recognised `TagPvidOnly` but the aggregator never handled it, so those ports fell through to routed.

ERS `TagPvidOnly` semantics are unusual: PVID egress is tagged, other VLANs egress untagged — and ERS allows multiple untagged-egress VLANs in this mode (IEEE 802.1Q forbids and NetBox can't represent). The safest NetBox-aligned mapping is the same shape as `TagAll`: trunk, no untagged native, every member VLAN in the tagged list. Folded the two modes into the same mapper branch (inherits the empty-membership → routed defensive guard).

New unit test pins the behaviour: `test_ers_tag_pvid_only_yields_trunk_no_native`.

All 969 tests pass, ruff clean.

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/ubiquiti_edgeswitch.py [hunk 1]
```diff
@@ -158,6 +160,401 @@ def _parse_vlan_members(config: str) -> dict[str, list[str]]:
...
+        if vid in entry["participation"]:
+            entry["participation"].remove(vid)
+
+
+def _trunk_allowed_add(entry: dict, spec: str) -> None:
+    """Append each parsed VID to tagging + participation (deduplicated)."""
+    for v in _expand_vlan_tokens(spec):
+        vid = int(v)
+        if vid not in entry["tagging"]:
+            entry["tagging"].append(vid)
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/ubiquiti_edgeswitch.py:313]: Fixed in 30fb866. Confirmed: a config with `switchport trunk native vlan 1` + `switchport trunk allowed vlan 1,10,20` was incorrectly stripping the native (VLAN 1) from `untagged_members` because it was also in `tagging`. Extracted `_normalise_edgesw_membership` that strips PVID from `tagging` before deriving `untagged_members` so the native ends up in the right bucket. New unit test `test_edgesw_cisco_trunk_native_vlan_excluded_from_tagged` pins the corrected output (`mode=trunk, native=1, tagged=[10,20]`). All 970 tests pass, ruff clean.

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/brocade_netiron.py [hunk 2]
```diff
@@ -387,3 +616,99 @@ def get_vlans(self) -> dict:
...
+          this :func:`apply_interface_vlans` would silently drop VLAN
+          mappings for VE/LAG/Loopback interfaces (Codex P1 #391).
+
+        Returns an empty dict on parse failure — callers fall back to
+        the bare IDs.
+        """
+        try:
+            raw = self.device.send_command("show interfaces")
+            parsed = parse_output(
+                platform="brocade_netiron", command="show interfaces", data=raw,
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/brocade_netiron.py:690]: Already fixed in 38caa77 (round-2 of this PR). The regex now allows optional leading digits via `(\d*[A-Za-z]+Ethernet)` so `10GigabitEthernet3/4` / `40GigabitEthernet5/1` enter the canonical map. Pinned by `test_netiron_canonical_map_handles_digit_leading_prefix_and_named_types`.

File: device-discovery/custom_napalm/avaya_ers.py [hunk 3]
```diff
@@ -138,6 +144,227 @@ def _mask_to_prefix(mask: str) -> int:
...
+        tokens.append(str(member))
+    return tokens
+
+
+def _expand_ers_port_chunk(part: str) -> list[str]:
+    """Expand a single comma-separated chunk of an ERS port-list."""
+    if not part or part.upper() in ("NONE", "ALL"):
+        return []
+    if "-" not in part:
+        return [part]
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/avaya_ers.py]: Fixed in 6733063. Confirmed: `_collect_ers_per_port_membership` treated `1/ALL` as a literal port name, so the trunk-mode aggregators saw empty `member_vids` and dropped real ports as routed. Extended `_expand_ers_port_chunk`/`_expand_ers_port_list` with a `known_ports` parameter (the catalog from `show vlan interface info`); chassis-wide `ALL` and unit-wide `<unit>/ALL` now resolve to concrete ports from the catalog. New unit test `test_ers_expand_port_list_wildcards` pins ALL/`<unit>/ALL`/empty-catalog behaviour.

File: device-discovery/custom_napalm/ubiquiti_edgeswitch.py [hunk 4]
```diff
@@ -158,6 +160,429 @@ def _parse_vlan_members(config: str) -> dict[str, list[str]]:
...
+    """
+    pvid = coerce_vid(membership.get("pvid"))
+    participation = [
+        v for v in membership.get("participation", []) if coerce_vid(v) is not None
+    ]
+    tagging = [v for v in membership.get("tagging", []) if coerce_vid(v) is not None]
+    if pvid is not None:
+        tagging = [v for v in tagging if v != pvid]
+    untagged_members = [v for v in participation if v not in tagging]
+    return participation, tagging, untagged_members
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/ubiquiti_edgeswitch.py:527]: Fixed in 6733063. Added `_dedupe_keep_order` and applied it to both `participation` and `tagging` in `_normalise_edgesw_membership` before deriving `untagged_members`. New unit test `test_edgesw_normalise_dedupes_repeated_vids` pins both the access-on-100 case (repeated `vlan participation include 100`) and the mixed native+Cisco case.

File: device-discovery/custom_napalm/ubiquiti_edgeswitch.py [hunk 5]
```diff
@@ -158,6 +160,429 @@ def _parse_vlan_members(config: str) -> dict[str, list[str]]:
...
+    * ``Trunk``   – tagged VIDs from ``vlan tagging``; native VID is the
+      untagged VID present in the membership list (typically VLAN 1 by
+      default), or None when no untagged member exists.
+    * ``General`` – collapsed to trunk: tagged from ``vlan tagging``, native
+      from the (single) untagged VID in the membership list.
+    * ``Routed`` or anything else → routed.
+
+    When membership data is missing (no running-config block for the port),
+    falls back to the PVID from the summary as a best-effort native/access
+    VID. Trunk with no membership data degrades to trunk no native.
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/ubiquiti_edgeswitch.py]: Docstring updated in 6733063 to match the current defensive behaviour. The PVID-fallback prose was stale (it described an earlier round-1 design that round-1 review tightened). Now explicitly documents that empty membership → routed for every mode, plus the new `tagged-all` (Cisco `switchport trunk allowed vlan all`) and `except` (Cisco `switchport trunk allowed vlan except`) paths.

File: device-discovery/custom_napalm/avaya_ers.py [hunk 6]
```diff
@@ -138,6 +144,227 @@ def _mask_to_prefix(mask: str) -> int:
...
+    """
+    Map per-port info + VLAN-membership list to ``SwitchportInfo``.
+
+    Tagging interpretation (ERS):
+      - ``UntagAll``      → access on PVID (port carries only PVID, untagged)
+      - ``UntagPvidOnly`` → trunk with native=PVID, tagged=members minus PVID
+      - ``TagAll``        → trunk, no native; tagged=members minus PVID
+                            (PVID is sent tagged, but in NetBox semantics there
+                            is no "untagged" → treat as trunk-no-native; PVID
+                            stays in the tagged list when membership reports it)
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/avaya_ers.py]: Docstring updated in 6733063 to accurately describe each tagging mode: TagAll → trunk-no-native + tagged=members (PVID stays in the list when reported); TagPvidOnly → same shape as TagAll; Hybrid → same as UntagPvidOnly; UntagPvidOnly with empty members → routed.

**Comment by leoparente** [comment]
@codex review

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/ubiquiti_edgeswitch.py [hunk 7]
```diff
@@ -158,6 +160,460 @@ def _parse_vlan_members(config: str) -> dict[str, list[str]]:
...
+# ---------------------------------------------------------------------------
+
+# Match physical and LAG interface blocks; skip ``interface vlan <id>`` SVIs.
+# EdgeSwitch CLI accepts both single-token (``interface 0/1``, ``interface lag1``)
+# and split-token (``interface lag 1``) forms — capture the rest of the line and
+# normalise whitespace on use so both reduce to the same key (``lag1``).
+_ES_INTF_BLOCK_RE = re.compile(r"^interface\s+(.+?)\s*$")
+_ES_VLAN_PVID_RE = re.compile(r"^vlan\s+pvid\s+(\d+)\s*$")
+_ES_VLAN_PART_RE = re.compile(r"^vlan\s+participation\s+include\s+(.+)$")
+_ES_VLAN_TAG_RE = re.compile(r"^vlan\s+tagging\s+(.+)$")
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/ubiquiti_edgeswitch.py:223]: Fixed in d039168. Added `_ES_VLAN_PART_EXCLUDE_RE` plus an exclude branch in `_apply_native_vlan_directive` that filters the VIDs out of both `participation` and `tagging`. New unit test `test_edgesw_native_vlan_participation_exclude_drops_vid` covers the include+exclude in the same block case.

File: device-discovery/custom_napalm/brocade_netiron.py [hunk 8]
```diff
@@ -387,3 +616,99 @@ def get_vlans(self) -> dict:
...
+        so LAG members declared as ``tagged lag <N>`` / ``untagged lag <N>``
+        are captured alongside physical ``ethe`` ports — heavily used on
+        NetIron chassis platforms.
+
+        Bare port IDs from the running-config (``1/1``, ``3/4``) are
+        canonicalised to the same speed-prefixed form ``get_interfaces()``
+        emits (``GigabitEthernet1/1``, ``10GigabitEthernet3/4``) using a
+        port→speed lookup from ``show interfaces brief`` — without this
+        ``apply_interface_vlans()`` would silently drop every entry due to
+        exact-name mismatch with the discovered Interface entities.
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/brocade_netiron.py:648]: Docstring fixed in d039168. The remap is built from `show interfaces` (not `show interfaces brief`) inside `_netiron_canonical_name_map` — each interface block is headed by the canonical speed-prefixed name, which is parsed via the brocade_netiron ntc-template. Doc now matches the code.

File: device-discovery/custom_napalm/ubiquiti_edgeswitch.py [hunk 9]
```diff
@@ -158,6 +160,460 @@ def _parse_vlan_members(config: str) -> dict[str, list[str]]:
...
+    """Drop each parsed VID from tagging + participation if present."""
+    for v in _expand_vlan_tokens(spec):
+        try:
+            vid = int(v)
+        except ValueError:
+            continue
+        if vid in entry["tagging"]:
+            entry["tagging"].remove(vid)
+        if vid in entry["participation"]:
+            entry["participation"].remove(vid)
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/ubiquiti_edgeswitch.py]: Fixed in d039168. Switched `_trunk_allowed_remove` from `list.remove(vid)` (which only drops the first occurrence) to list-comprehension filtering using a set of excluded VIDs, so every occurrence is purged. New unit test `test_edgesw_trunk_allowed_remove_purges_duplicates` pins the behaviour with a duplicated VID 10 in both participation and tagging.

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/brocade_netiron.py [hunk 10]
```diff
@@ -387,3 +616,101 @@ def get_vlans(self) -> dict:
...
+            raw = self.device.send_command("show interfaces")
+            parsed = parse_output(
+                platform="brocade_netiron", command="show interfaces", data=raw,
+            )
+        except Exception:
+            logger.debug("NetIron show interfaces failed for canonical map", exc_info=True)
+            return {}
+        # Speed-prefixed Ethernet: prefix may start with digits (10/40/100/400),
+        # so allow optional leading digits.
+        ethernet_re = re.compile(r"^(\d*[A-Za-z]+Ethernet)(\d+/\d+(?::\d+)?)$")
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/brocade_netiron.py]: Fixed in 02449d1. Suffix is now `\d+(?:/\d+(?::\d+)?)?` so both stacked forms (`1/1`, `1/2:1`) and bare-digit CES forms (`1`, `11`) enter the canonical map. New unit test `test_netiron_canonical_map_handles_bare_digit_ces_form` pins the bare-numeric path with `GigabitEthernet1`, `GigabitEthernet11`, and `10GigabitEthernet3`.

File: device-discovery/custom_napalm/avaya_ers.py [hunk 11]
```diff
@@ -138,6 +144,276 @@ def _mask_to_prefix(mask: str) -> int:
...
+
+    Newer firmware emits a ``PID`` column inline (``vlan_pid``); older firmware
+    emits a ``Port Members:`` continuation line captured into the List value
+    ``vlan_port_members``. We process both — duplicates are de-duplicated by the
+    caller's per-VLAN membership-set.
+    """
+    tokens: list[str] = []
+    pid = row.get("vlan_pid", "")
+    if pid:
+        tokens.append(str(pid))
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/avaya_ers.py]: Fixed in 02449d1. Confirmed: the avaya_ers ntc-template captures `VLAN_PID` from the **PID** (Protocol ID) column — a hex string like `0x0000`, not a port identifier. Dropped `vlan_pid` from `_ers_row_port_tokens`; only the actual `vlan_port_members` (List, captured from `Port Members:` continuation lines) is now used. New unit test `test_ers_row_port_tokens_ignores_protocol_id` pins the behaviour.

File: device-discovery/custom_napalm/ubiquiti_edgeswitch.py [hunk 12]
```diff
@@ -158,6 +160,482 @@ def _parse_vlan_members(config: str) -> dict[str, list[str]]:
...
+    if op == "except":
+        entry["allowed_except"] = True
+        return
+    if spec.strip().lower() == "all":
+        entry["allowed_all"] = True
+        return
+    if op == "remove":
+        _trunk_allowed_remove(entry, spec)
+    else:
+        _trunk_allowed_add(entry, spec)
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/ubiquiti_edgeswitch.py:356]: Fixed in 02449d1. After `allowed_all=True`, any subsequent `add` / `remove` directive now flips to the `allowed_except` routed fallback (the same path the explicit `except` keyword takes) and clears `allowed_all`, since the effective set is `all ± some` which NetBox cant represent without enumerating the chassis. Two new unit tests pin both branches: `test_edgesw_allowed_all_then_remove_falls_back_to_routed` and `test_edgesw_allowed_all_then_add_falls_back_to_routed`.

File: device-discovery/tests/custom_drivers/brocade_netiron/mock_data/test_get_interfaces_vlans/trunk_native/expected_result.json [hunk 13]
```diff
@@ -0,0 +1,10 @@
+{
+  "1/1": {
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/tests/custom_drivers/brocade_netiron/mock_data/test_get_interfaces_vlans/trunk_native/expected_result.json:2]: Added in 02449d1. New scenario `with_canonical_remap/` provides both `show running-config_vlan.txt` (referencing `1/1`, `1/2`, `3/4`, `lag 5`) and `show interfaces.txt` (with canonical-name headers `GigabitEthernet1/1`, `10GigabitEthernet3/4`, `Lag5`, etc.). The expected_result.json asserts the canonical names so the end-to-end remap is validated. Existing scenarios stay as bare-key fallback tests (no behaviour regression).

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/ubiquiti_edgeswitch.py [hunk 14]
```diff
@@ -158,6 +160,496 @@ def _parse_vlan_members(config: str) -> dict[str, list[str]]:
...
+    every mode falls back to routed rather than guessing — apply_interface_vlans()
+    would otherwise PATCH NetBox with a guessed VID and clobber the
+    existing untagged_vlan / tagged_vlans. Mirrors the dell_powerconnect
+    batch-4 tightening from #390.
+    """
+    del port  # unused; kept for parity with PowerConnect signature
+    mode_raw = (summary.get("mode") or "").lower()
+
+    if membership is None:
+        membership = {"participation": [], "tagging": [], "pvid": None}
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/ubiquiti_edgeswitch.py:580]: Fixed in d3873e6. EdgeSwitch `show running-config` only emits configured interfaces, so default-config Access ports appear in the switchport summary but have no running-config block. The summary is authoritative for mode + PVID, so we now trust it for Access mode specifically when membership is unambiguously absent (no participation, no tagging). Membership present but mismatched still routes (preserves the round-1 tightening). New tests `test_edgesw_access_no_membership_trusts_summary_pvid` and `test_edgesw_access_no_membership_no_pvid_yields_routed` pin both branches.

File: device-discovery/custom_napalm/ubiquiti_edgeswitch.py [hunk 15]
```diff
@@ -158,6 +160,496 @@ def _parse_vlan_members(config: str) -> dict[str, list[str]]:
...
+            native_vlan=native_vid,
+            allowed_vlans=tagging if tagging else None,
+        )
+
+    if mode_raw == "general":
+        # General mode: collapse to trunk. Native = untagged member;
+        # tagged = explicit tagging list.
+        native_vid = untagged_members[0] if untagged_members else None
+        if native_vid is None and not tagging:
+            return _edgesw_routed()
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/ubiquiti_edgeswitch.py]: Fixed in d3873e6. Trunk/General modes now route when `len(untagged_members) > 1` instead of silently picking `untagged_members[0]` and dropping the rest. Matches the multi-untagged routing in netiron/slx/unifiswitch/dell_powerconnect. New unit test `test_edgesw_trunk_multi_untagged_yields_routed` pins the behaviour.

File: device-discovery/custom_napalm/avaya_ers.py [hunk 16]
```diff
@@ -138,6 +144,273 @@ def _mask_to_prefix(mask: str) -> int:
...
+
+    if tagging == "untagall":
+        return SwitchportInfo(
+            enabled=True,
+            admin_mode="access",
+            oper_mode="access",
+            access_vlan=pvid,
+            native_vlan=None,
+            allowed_vlans=None,
+        )
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/avaya_ers.py:259]: Fixed in d3873e6. Added a defensive guard: UntagAll with `pvid is None` (missing or out-of-range PVID — `coerce_vid` rejects values outside 1..4094) → routed instead of emitting `access_vlan=None` which would clobber the existing NetBox untagged_vlan via PATCH. New unit test `test_ers_untag_all_with_invalid_pvid_yields_routed` covers both None and out-of-range cases.

**Comment by leoparente** [comment]
@codex review
